src/ids-host.test.ts

```diff
@@ -0,0 +1,326 @@
+      recordToUrl({
+        name: "hi",
+        port: 123,
+        priority: 1,
+        weight: 1,
+      }),
+    ).toStrictEqual(new URL("https://hi:123"));
+  });
+
+  test("an invalid record", () => {
+    expect(
+      recordToUrl({
+        name: "!",
+        port: 99999999999,
+        priority: 1,
+        weight: 1,
+      }),
+    ).toStrictEqual(undefined);
+  });
+});
+
+describe("discoverServicesStub", async () => {
+  type TestCase = {
+    description: string;
+    lookup: () => Promise<SrvRecord[]>;
+    timeout?: number;
+    expected: SrvRecord[];
+  };
+
+  const testCases: TestCase[] = [
+    {
+      description: "lookup has an exception",
+      lookup: async () => {
+        throw new Error("oops");
+      },
+      expected: [],
+      timeout: 0,
+    },
+
+    {
+      description: "basic in / out",
+      lookup: async () => {
+        return mkPromise(() => {
+          return [mkRecord(1)];
+        });
+      },
+      expected: [mkRecord(1)],
+    },
+
+    {
+      description: "Records with invalid suffixes are omitted",
+      lookup: async () => {
+        return mkPromise(() => {
+          return [
+            mkRecord(1, 1, "mallory.com"),
+            mkRecord(1, 1, "install.detsys.dev"),
+            mkRecord(1, 1, "install.determinate.systems"),
+          ];
+        });
+      },
+      expected: [
+        mkRecord(1, 1, "install.detsys.dev"),
+        mkRecord(1, 1, "install.determinate.systems"),
+      ],
+    },
+
+    {
+      description: "lookup loses the race",
+      lookup: async () => {
+        return new Promise((r) => setTimeout(r, 1000, [mkRecord(123)]));
+      },
+      expected: [],
+      timeout: 100,
+    },
+    {
+      description: "lookup wins the race",
+      lookup: async () => {
+        return new Promise((r) => setTimeout(r, 100, [mkRecord(456)]));
+      },
+      expected: [mkRecord(456)],
+      timeout: 1000,
+    },
+  ];
+
+  for (const { description, lookup, timeout, expected } of testCases) {
+    test(description, async () => {
+      const ret = await discoverServicesStub(lookup(), timeout || 0);
+      expect(ret).toStrictEqual(expected);
+    });
+  }
+});
+
+test("orderRecordsByPriorityWeight does that", () => {
+  expect(
+    orderRecordsByPriorityWeight([
+      mkRecord(3, 3),
+      mkRecord(1000, 1),
+      mkRecord(2, 2),
+      mkRecord(1, 1),
+    ]),
+  ).toStrictEqual([
+    mkRecord(1000, 1),
+    mkRecord(1, 1),
+    mkRecord(2, 2),
+    mkRecord(3, 3),
+  ]);
+});
+
+test("weightedRandom handles empty and single-element records", () => {
+  expect(weightedRandom([]), "one element passes through").toStrictEqual([]);
+
+  expect(
+    weightedRandom([mkRecord(1)]),
+    "empty lists aren't crashing",
+  ).toStrictEqual([mkRecord(1)]);
+});
+
+test("weightedRandom orders records acceptably predictably", () => {
+  const records: SrvRecord[] = [mkRecord(1), mkRecord(2), mkRecord(3)];
+
+  const counts: Map<number, number> = new Map();
```

Comment:
I suspect there's a way to do this without needing a bunch of lint ignores. I'll look into it.